private/bufpkg/bufplugin/bufplugindocker/context.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufplugindocker
+
+import (
+	"archive/tar"
+	"bytes"
+	"io"
+	"io/fs"
+	"time"
+)
+
+// createDockerContext builds a light-weight Docker context (tar file) from a given Dockerfile.
+// This is used to create images that don't rely on any other filesystem state to copy into the context.
+func createDockerContext(dockerfile io.Reader) (io.Reader, error) {
+	dockerfileBytes, err := io.ReadAll(dockerfile)
+	if err != nil {
+		return nil, err
+	}
+	var buffer bytes.Buffer
+	writer := tar.NewWriter(&buffer)
+	dockerfileInfo := &fileInfo{
+		name: "Dockerfile",
```

Comment:
We store the file as `Dockerfile` even if it was named `Dockerfile.plugin` - this just removes one more non-default setting we have to pass to the Docker build.

Reply:
I wonder if we should just make the default filename we look for `Dockerfile`. As long as it's in a single directory alongside the `buf.plugin.yaml`, it would be obvious what it represents. With that said, I like the `.plugin` suffix so that it's clear what `Dockerfile` will be used as the primary entrypoint (if there's ever a reason to have multiple Dockerfiles exists alongside each other).

What are your thoughts?

Reply:
I like the plugin suffix. I'm only storing it as Dockerfile in the context since we don't have to set it in two places (here and in ImageBuildOptions.Dockerfile).

Reply:
Reimplementing os.FileInfo is required here to create a tar entry from an io.Reader. This code is written where we don't make assumptions on where the Dockerfile exists (is it loaded from a tar file, a zip file, a directory, etc.).

What we're basically doing here is the following:
* Take Dockerfile.plugin from a storage.ReadBucket (this could be any of the supported types).
* Create a tar file which writes a single file - including the entry (this is what the FileInfo is needed for) and the contents of the file.
* Pass that as input to the Docker image build command. If we specify the file in the tar file as Dockerfile, then we don't have to override the default value here: https://docs.docker.com/engine/api/v1.41/#operation/ImageBuild.

We could change this to ensure that we have a real directory to build from (essentially extracting the Dockerfile.plugin to a temporary directory if it is loaded from an archive file), but that felt like more edge cases to cover with the initial pass.

Reply:
Got it, this seems reasonable for now then. Please write up some of the justification on the comment above the `fileInfo` type declaration though - this context is important for future readers.